File: packages/less/src/less/tree/element.test.js
import { describe, it, expect, vi } from 'vitest';
import Element from './element';
import Node from './node';
import Paren from './paren';
import Combinator from './combinator';

describe('Element', () => {
    describe('constructor', () => {
        it('should create Element with string combinator', () => {
            const element = new Element('>', 'div', false);
            expect(element.combinator).toBeInstanceOf(Combinator);
            expect(element.combinator.value).toBe('>');
            expect(element.value).toBe('div');
            expect(element.isVariable).toBe(false);
        });

        it('should create Element with Combinator instance', () => {
            const combinator = new Combinator('+');
            const element = new Element(combinator, 'span', false);
            expect(element.combinator).toBe(combinator);
            expect(element.value).toBe('span');
        });

        it('should trim string values', () => {
            const element = new Element('>', '  p  ', false);
            expect(element.value).toBe('p');
        });

        it('should handle empty value', () => {
            const element = new Element('>', null, false);
            expect(element.value).toBe('');
        });

        it('should handle non-string values', () => {
            const obj = { type: 'test' };
            const element = new Element('>', obj, false);
            expect(element.value).toBe(obj);
        });

        it('should set isVariable flag', () => {
            const element = new Element('>', 'var', true);
            expect(element.isVariable).toBe(true);
        });

        it('should inherit from Node', () => {
            const element = new Element('>', 'div', false);
            expect(element).toBeInstanceOf(Node);
        });

        it('should handle undefined value', () => {
            const element = new Element('>', undefined, false);
            expect(element.value).toBe('');
        });

        it('should handle empty string value', () => {
            const element = new Element('>', '', false);
            expect(element.value).toBe('');
        });

        it('should handle space combinator specially', () => {
            const element = new Element(' ', 'div', false);
            expect(element.combinator.value).toBe(' ');
            expect(element.combinator.emptyOrWhitespace).toBe(true);
        });

        it('should store file info and index', () => {
            const fileInfo = { filename: 'test.less' };
            const element = new Element('>', 'div', false, 42, fileInfo);
            expect(element._fileInfo).toBe(fileInfo);
            expect(element._index).toBe(42);
        });

        it('should copy visibility info', () => {
            const visInfo = { visibilityBlocks: 1, nodeVisible: true };
            const element = new Element('>', 'div', false, 1, {}, visInfo);
            expect(element.visibilityInfo()).toEqual(visInfo);
        });

        it('should handle malformed fileInfo', () => {
            const element = new Element('>', 'div', false, 1, null);
            expect(element._fileInfo).toBeNull();

            const element2 = new Element('>', 'div', false, 1, undefined);
            expect(element2._fileInfo).toBeUndefined();
        });

        it('should handle malformed visibilityInfo', () => {
            const element = new Element('>', 'div', false, 1, {}, null);
            expect(element.visibilityInfo()).toEqual({});

            const element2 = new Element(
                '>',
                'div',
                false,
                1,
                {},
                { visibilityBlocks: 'invalid' }
            );
            expect(element2.visibilityInfo()).toEqual({
                visibilityBlocks: 'invalid'
            });
        });

        it('should handle invalid index', () => {
            const element = new Element('>', 'div', false, 'not-a-number');
            expect(element._index).toBe('not-a-number');
        });
    });

    describe('accept', () => {
        it('should visit combinator', () => {
            const element = new Element('>', 'div', false);
            const visitor = {
                visit: () => new Combinator('+')
            };
            element.accept(visitor);
            expect(element.combinator.value).toBe('+');
        });

        it('should visit object values', () => {
            const value = { type: 'test' };
            const element = new Element('>', value, false);
            const visitor = {
                visit: () => ({ type: 'visited' })
            };
            element.accept(visitor);
            expect(element.value.type).toBe('visited');
        });

        it('should not visit string values', () => {
            const element = new Element('>', 'div', false);
            const visitor = {
                visit: () => 'visited'
            };
            element.accept(visitor);
            expect(element.value).toBe('div');
        });

        it('should handle undefined value', () => {
            const element = new Element('>', undefined, false);
            const visitor = {
                visit: () => 'visited'
            };
            element.accept(visitor);
            expect(element.value).toBe('');
        });

        it('should handle null value', () => {
            const element = new Element('>', null, false);
            const visitor = {
                visit: () => 'visited'
            };
            element.accept(visitor);
            expect(element.value).toBe('');
        });

        it('should handle undefined/null values', () => {
            const element = new Element('>', undefined, false);
            const visitor = {
                visit: () => 'visited'
            };
            element.accept(visitor);
            expect(element.value).toBe('');
        });
    });

    describe('eval', () => {
        it('should create new Element with same properties', () => {
            const element = new Element('>', 'div', false, 1, {
                filename: 'test.less'
            });
            const result = element.eval({});
            expect(result).toBeInstanceOf(Element);
            expect(result).not.toBe(element);
            expect(result.value).toBe('div');
            expect(result.isVariable).toBe(false);
            expect(result.getIndex()).toBe(1);
        });

        it('should eval value if it has eval method', () => {
            const value = {
                eval: () => 'evaluated'
            };
            const element = new Element('>', value, false);
            const result = element.eval({});
            expect(result.value).toBe('evaluated');
        });

        it('should handle eval returning null', () => {
            const value = {
                eval: () => null
            };
            const element = new Element('>', value, false);
            const result = element.eval({});
            expect(result.value).toBe('');
        });

        it('should handle value without eval method', () => {
            const element = new Element('>', 'div', false);
            const result = element.eval({});
            expect(result.value).toBe('div');
        });

        it('should preserve visibility info during eval', () => {
            const visInfo = { visibilityBlocks: 1, nodeVisible: true };
            const element = new Element('>', 'div', false, 1, {}, visInfo);
            const result = element.eval({});
            expect(result.visibilityInfo()).toEqual(visInfo);
        });

        it('should preserve file info during eval', () => {
            const fileInfo = { filename: 'test.less' };
            const element = new Element('>', 'div', false, 1, fileInfo);
            const result = element.eval({});
            expect(result.fileInfo()).toEqual(fileInfo);
        });

        it('should handle value.eval throwing error', () => {
            const value = {
                eval: () => {
                    throw new Error('eval error');
                }
            };
            const element = new Element('>', value, false);
            expect(() => element.eval({})).toThrow('eval error');
        });

        it('should handle value.eval returning unexpected types', () => {
            const tests = [
                { evalReturn: undefined, expected: '' },
                { evalReturn: () => {}, expectedType: 'function' },
                {
                    evalReturn: Symbol('test'),
                    expectedType: 'symbol',
                    expectedDescription: 'test'
                },
                { evalReturn: 42, expected: 42 },
                { evalReturn: false, expected: '' }
            ];

            tests.forEach(
                ({
                    evalReturn,
                    expected,
                    expectedType,
                    expectedDescription
                }) => {
                    const value = {
                        eval: () => evalReturn
                    };
                    const element = new Element('>', value, false);
                    const result = element.eval({});

                    if (expectedType === 'function') {
                        expect(typeof result.value).toBe('function');
                    } else if (expectedType === 'symbol') {
                        expect(typeof result.value).toBe('symbol');
                        expect(result.value.description).toBe(
                            expectedDescription
                        );
                    } else {
                        expect(result.value).toBe(expected);
                    }
                }
            );
        });
    });

    describe('clone', () => {
        it('should create exact copy with all properties', () => {
            const fileInfo = { filename: 'test.less' };
            const visInfo = { visibilityBlocks: 1, nodeVisible: true };
            const element = new Element('>', 'div', true, 1, fileInfo);
            element.copyVisibilityInfo(visInfo);

            const clone = element.clone();
            expect(clone).toBeInstanceOf(Element);
            expect(clone).not.toBe(element);
            expect(clone.combinator.value).toBe(element.combinator.value);
            expect(clone.value).toBe(element.value);
            expect(clone.isVariable).toBe(element.isVariable);
            expect(clone.getIndex()).toBe(element.getIndex());
            expect(clone.fileInfo()).toEqual(element.fileInfo());
            expect(clone.visibilityInfo()).toEqual(element.visibilityInfo());
        });
    });

    describe('toCSS', () => {
        it('should combine combinator and value CSS', () => {
            const element = new Element('>', 'div', false);
            expect(element.toCSS()).toBe(' > div');
        });

        it('should handle Paren values', () => {
            const parenValue = {
                genCSS: (context, output) => output.add('test')
            };
            const paren = new Paren(parenValue);
            const element = new Element('>', paren, false);
            const result = element.toCSS();
            expect(result).toBe(' > (test)');
        });

        it('should return empty string for empty value with & combinator', () => {
            const element = new Element('&', '', false);
            expect(element.toCSS()).toBe('');
        });

        it('should handle values with toCSS method', () => {
            const value = { toCSS: () => 'custom' };
            const element = new Element('>', value, false);
            expect(element.toCSS()).toBe(' > custom');
        });

        it('should handle firstSelector context with Paren value', () => {
            const mockContext = { firstSelector: false };
            const parenValue = {
                genCSS: (ctx, output) => output.add('test'),
                toCSS: (ctx) => {
                    expect(ctx.firstSelector).toBe(true);
                    return 'test';
                }
            };
            const paren = new Paren(parenValue);
            const element = new Element('>', paren, false);
            element.toCSS(mockContext);
            expect(mockContext.firstSelector).toBe(false);
        });

        it('should handle different combinator types', () => {
            const tests = [
                { combinator: ' ', expected: ' div' },
                { combinator: '|', expected: '|div' },
                { combinator: '>', expected: ' > div' },
                { combinator: '+', expected: ' + div' },
                { combinator: '~', expected: ' ~ div' }
            ];

            tests.forEach(({ combinator, expected }) => {
                const element = new Element(combinator, 'div', false);
                expect(element.toCSS()).toBe(expected);
            });
        });

        it('should handle empty values', () => {
            const element = new Element('>', '', false);
            expect(element.toCSS()).toBe(' > ');
        });

        it('should handle & combinator with empty value', () => {
            const element = new Element('&', '', false);
            expect(element.toCSS()).toBe('');
        });

        it('should handle value.toCSS throwing error', () => {
            const value = {
                toCSS: () => {
                    throw new Error('toCSS error');
                }
            };
            const element = new Element('>', value, false);
            expect(() => element.toCSS()).toThrow('toCSS error');
        });

        it('should handle value.toCSS returning unexpected types', () => {
            const tests = [
                { cssReturn: undefined, expected: ' > undefined' },
                { cssReturn: null, expected: ' > null' },
                { cssReturn: 42, expected: ' > 42' },
                { cssReturn: false, expected: ' > false' },
                { cssReturn: {}, expected: ' > [object Object]' }
            ];

            tests.forEach(({ cssReturn, expected }) => {
                const value = {
                    toCSS: () => cssReturn
                };
                const element = new Element('>', value, false);
                expect(element.toCSS()).toBe(expected);
            });
        });

        it('should handle Paren value with throwing methods', () => {
            const parenValue = {
                genCSS: () => {
                    throw new Error('genCSS error');
                }
            };
            const paren = new Paren(parenValue);
            const element = new Element('>', paren, false);
            expect(() => element.toCSS()).toThrow('genCSS error');
        });
    });

    describe('genCSS', () => {
        it('should add CSS with file info and index', () => {
            const fileInfo = { filename: 'test.less' };
            const element = new Element('>', 'div', false, 42, fileInfo);
            const output = {
                add: vi.fn()
            };
            element.genCSS({}, output);
            expect(output.add).toHaveBeenCalledWith(' > div', fileInfo, 42);
        });

        it('should handle output.add throwing error', () => {
            const element = new Element('>', 'div', false);
            const output = {
                add: () => {
                    throw new Error('add error');
                }
            };
            expect(() => element.genCSS({}, output)).toThrow('add error');
        });

        it('should handle null/undefined output', () => {
            const element = new Element('>', 'div', false);
            expect(() => element.genCSS({}, null)).toThrow();
            expect(() => element.genCSS({}, undefined)).toThrow();
        });
    });

    describe('edge cases', () => {
        it('should handle circular references', () => {
            const obj = {};
            Object.assign(obj, { self: obj });
            const element = new Element('>', obj, false);
            expect(() => element.toCSS()).not.toThrow();
        });

        it('should handle very long values', () => {
            const longString = 'a'.repeat(10000);
            const element = new Element('>', longString, false);
            expect(() => element.toCSS()).not.toThrow();
            expect(element.toCSS()).toBe(` > ${longString}`);
        });

        it('should handle special characters in values', () => {
            const specialChars = '!@#$%^&*()_+{}[]|\\:;"\'<>,.?/~`';
            const element = new Element('>', specialChars, false);
            expect(() => element.toCSS()).not.toThrow();
            expect(element.toCSS()).toBe(` > ${specialChars}`);
        });

        it('should handle emoji and unicode characters', () => {
            const unicodeChars = '🌟 你好 привет مرحبا';
            const element = new Element('>', unicodeChars, false);
            expect(() => element.toCSS()).not.toThrow();
            expect(element.toCSS()).toBe(` > ${unicodeChars}`);
        });
    });
});

File: packages/less/src/less/go_parser/element_test.go
package go_parser

import (
	"reflect"
	"strings"
	"testing"
)

func TestElement(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("should create Element with string combinator", func(t *testing.T) {
			element := NewElement(">", "div", false, 0, nil, nil)
			if element.Combinator == nil {
				t.Error("Expected combinator to not be nil")
			}
			if element.Combinator.Value != ">" {
				t.Errorf("Expected combinator value to be '>', got '%s'", element.Combinator.Value)
			}
			if element.Value != "div" {
				t.Errorf("Expected value to be 'div', got '%v'", element.Value)
			}
			if element.IsVariable {
				t.Error("Expected isVariable to be false")
			}
		})

		t.Run("should create Element with Combinator instance", func(t *testing.T) {
			combinator := NewCombinator("+")
			element := NewElement(combinator, "span", false, 0, nil, nil)
			if element.Combinator != combinator {
				t.Error("Expected combinator to be the same instance")
			}
			if element.Value != "span" {
				t.Errorf("Expected value to be 'span', got '%v'", element.Value)
			}
		})

		t.Run("should trim string values", func(t *testing.T) {
			element := NewElement(">", "  p  ", false, 0, nil, nil)
			if element.Value != "p" {
				t.Errorf("Expected value to be 'p', got '%v'", element.Value)
			}
		})

		t.Run("should handle empty value", func(t *testing.T) {
			element := NewElement(">", nil, false, 0, nil, nil)
			if element.Value != "" {
				t.Errorf("Expected value to be empty string, got '%v'", element.Value)
			}
		})

		t.Run("should handle non-string values", func(t *testing.T) {
			obj := map[string]string{"type": "test"}
			element := NewElement(">", obj, false, 0, nil, nil)
			if !reflect.DeepEqual(element.Value, obj) {
				t.Error("Expected value to be the same object")
			}
		})
	})

	t.Run("accept", func(t *testing.T) {
		t.Run("should visit combinator", func(t *testing.T) {
			element := NewElement(">", "div", false, 0, nil, nil)
			visitor := &visitorImpl{returnValue: NewCombinator("+")}
			element.Accept(visitor)
			if element.Combinator.Value != "+" {
				t.Errorf("Expected combinator value to be '+', got '%s'", element.Combinator.Value)
			}
		})

		t.Run("should visit object values", func(t *testing.T) {
			value := map[string]string{"type": "test"}
			element := NewElement(">", value, false, 0, nil, nil)
			visitor := &visitorImpl{returnValue: map[string]string{"type": "visited"}}
			element.Accept(visitor)
			if val, ok := element.Value.(map[string]string); !ok || val["type"] != "visited" {
				t.Error("Expected value to be updated by visitor")
			}
		})

		t.Run("should not visit string values", func(t *testing.T) {
			element := NewElement(">", "div", false, 0, nil, nil)
			visitor := &visitorImpl{returnValue: "visited"}
			element.Accept(visitor)
			if element.Value != "div" {
				t.Error("Expected string value to remain unchanged")
			}
		})
	})

	t.Run("eval", func(t *testing.T) {
		t.Run("should create new Element with same properties", func(t *testing.T) {
			fileInfo := map[string]any{"filename": "test.less"}
			element := NewElement(">", "div", false, 1, fileInfo, nil)
			result := element.Eval(map[string]any{})

			if result == element {
				t.Error("Expected result to be a new instance")
			}
			if result.Value != "div" {
				t.Error("Expected value to be preserved")
			}
			if result.IsVariable {
				t.Error("Expected isVariable to be preserved")
			}
			if result.GetIndex() != 1 {
				t.Error("Expected index to be preserved")
			}
		})

		t.Run("should eval value if it has eval method", func(t *testing.T) {
			value := &testEvaluator{returnValue: "evaluated"}
			element := NewElement(">", value, false, 0, nil, nil)
			result := element.Eval(map[string]any{})
			if result.Value != "evaluated" {
				t.Error("Expected value to be evaluated")
			}
		})
	})

	t.Run("toCSS", func(t *testing.T) {
		t.Run("should combine combinator and value CSS", func(t *testing.T) {
			element := NewElement(">", "div", false, 0, nil, nil)
			css := element.ToCSS(nil)
			if css != " > div" {
				t.Errorf("Expected CSS to be ' > div', got '%s'", css)
			}
		})

		t.Run("should handle Paren values", func(t *testing.T) {
			parenValue := &testCSSGenerator{cssValue: "test"}
			paren := NewParen(parenValue)
			element := NewElement(">", paren, false, 0, nil, nil)
			result := element.ToCSS(nil)
			if result != " > test" {
				t.Errorf("Expected CSS to be ' > test', got '%s'", result)
			}
		})

		t.Run("should return empty string for empty value with & combinator", func(t *testing.T) {
			element := NewElement("&", "", false, 0, nil, nil)
			if css := element.ToCSS(nil); css != "" {
				t.Errorf("Expected empty string, got '%s'", css)
			}
		})
	})

	t.Run("edge cases", func(t *testing.T) {
		t.Run("should handle malformed fileInfo", func(t *testing.T) {
			element := NewElement(">", "div", false, 1, nil, nil)
			if element.FileInfo() == nil {
				t.Error("Expected empty map for nil fileInfo, got nil")
			}

			element2 := NewElement(">", "div", false, 1, map[string]any{}, nil)
			if len(element2.FileInfo()) != 0 {
				t.Error("Expected empty map for empty fileInfo")
			}

			element3 := NewElement(">", "div", false, 1, map[string]any{"invalid": nil}, nil)
			if element3.FileInfo()["invalid"] != nil {
				t.Error("Expected nil value to be preserved in fileInfo")
			}
		})

		t.Run("should handle malformed visibilityInfo", func(t *testing.T) {
			element := NewElement(">", "div", false, 1, nil, nil)
			if element.VisibilityInfo() == nil {
				t.Error("Expected non-nil visibilityInfo")
			}

			element2 := NewElement(">", "div", false, 1, nil, map[string]any{
				"visibilityBlocks": "invalid",
				"nodeVisible":      123,
			})
			visInfo := element2.VisibilityInfo()
			if visInfo == nil {
				t.Error("Expected non-nil visibilityInfo even with invalid data")
			}
		})

		t.Run("should handle invalid indices", func(t *testing.T) {
			element := NewElement(">", "div", false, -1, nil, nil)
			if element.GetIndex() != -1 {
				t.Error("Expected negative index to be preserved")
			}

			element2 := NewElement(">", "div", false, 0, nil, nil)
			if element2.GetIndex() != 0 {
				t.Error("Expected zero index to be preserved")
			}

			maxInt := int(^uint(0) >> 1)
			element3 := NewElement(">", "div", false, maxInt, nil, nil)
			if element3.GetIndex() != maxInt {
				t.Error("Expected max int index to be preserved")
			}
		})

		t.Run("should handle circular references", func(t *testing.T) {
			// Create a circular reference in the value
			type circular struct {
				self *circular
			}
			c := &circular{}
			c.self = c

			element := NewElement(">", c, false, 0, nil, nil)
			css := element.ToCSS(nil)
			if css == "" {
				t.Error("Expected non-empty CSS output even with circular reference")
			}
		})

		t.Run("should handle very long values", func(t *testing.T) {
			longString := strings.Repeat("a", 10000)
			element := NewElement(">", longString, false, 0, nil, nil)
			css := element.ToCSS(nil)
			expected := " > " + longString
			if css != expected {
				t.Errorf("Expected CSS to be '%s', got '%s'", expected, css)
			}
		})

		t.Run("should handle special characters", func(t *testing.T) {
			tests := []struct {
				name     string
				value    string
				expected string
			}{
				{
					name:     "special chars",
					value:    "!@#$%^&*()_+{}[]|\\:;\"'<>,.?/~`",
					expected: " > !@#$%^&*()_+{}[]|\\:;\"'<>,.?/~`",
				},
				{
					name:     "emoji",
					value:    "🌟 ⭐️ 🌙",
					expected: " > 🌟 ⭐️ 🌙",
				},
				{
					name:     "unicode",
					value:    "你好 привет مرحبا",
					expected: " > 你好 привет مرحبا",
				},
				{
					name:     "zero width chars",
					value:    "a\u200Bb\u200Cc",
					expected: " > a\u200Bb\u200Cc",
				},
			}

			for _, tt := range tests {
				t.Run(tt.name, func(t *testing.T) {
					element := NewElement(">", tt.value, false, 0, nil, nil)
					css := element.ToCSS(nil)
					if css != tt.expected {
						t.Errorf("Expected CSS to be '%s', got '%s'", tt.expected, css)
					}
				})
			}
		})
	})
}

// Test helpers
type visitorImpl struct {
	returnValue any
}

func (v *visitorImpl) Visit(node any) any {
	return v.returnValue
}

type testEvaluator struct {
	returnValue any
}

func (e *testEvaluator) Eval(context any) any {
	return e.returnValue
}

type testCSSGenerator struct {
	cssValue string
}

func (g *testCSSGenerator) ToCSS(context any) string {
	return g.cssValue
} 